oci/auth.py

```diff
@@ -133,3 +136,90 @@ def lookup_credentials(
         return valid_cfgs[0].credentials
 
     return lookup_credentials
+
+
+def docker_credentials_lookup(
+    docker_cfg: str | None=None,
+    absent_ok: bool=False,
+) -> typing.Callable[[image_reference, Privileges, bool], OciConfig]:
+    '''
+    returns a credentials-lookup backed by docker's auth-config. By design, docker's auth-config
+    only allows configuring credentials per hostname. By default' docker-cfg is expected at
+    `$HOME/.docker/config.json`. Location of docker-cfg can be customised via docker_cfg parameter.
+
+    if no docker-cfg is found, raises RuntimeError, unless absent_ok is truthy, in which case the
+    returned lookup will never return any credentials (which might still be useful for readonly
+    operations that for many registries allow anonymous access).
+
+    Note that docker does not offer to configure credentials by permissions, too (hence privileges
+    parameter will be ignored)
+    '''
+    if not docker_cfg:
+        docker_cfg = os.path.join(os.environ.get('HOME', ''), '.docker/config.json')
+
+    if not os.path.isfile(docker_cfg):
+        if not absent_ok:
+            raise RuntimeError(f'not an existing file: {docker_cfg=}')
+
+        def find_nothing_lookup(
+            image_reference: str,
+            privileges: Privileges=Privileges.READONLY,
+            absent_ok: bool=False,
+        ):
+            if not absent_ok:
+                raise ValueError(f'no auth-cfg found in {docker_cfg=} for {image_reference=}')
+            return None
+
+        return find_nothing_lookup
+
+    def docker_auth_lookup(
+        image_reference: str,
+        privileges: Privileges=Privileges.READONLY,
+        absent_ok: bool=False,
+    ):
+        # re-read docker-cfg to reflect fs-updates
+        with open(docker_cfg) as f:
+            docker_auth = json.load(f)
+            auths = docker_auth.get('auths', None)
+
+        if not auths:
+            # docker-cfg might be empty - do not handle as an error; however, we can never serve
+            # anything useful
+            if not absent_ok:
+                raise ValueError(f'no auth-cfg found in {docker_cfg=} for {image_reference=}')
+            return None
+
+        if image_reference.startswith('/'):
+            # if it is a relative reference, we have not means to find appropriate cfg.
+            if not absent_ok:
+                raise ValueError(f'no auth-cfg found in {docker_cfg=} for {image_reference=}')
+            return None
+
+        # ignore ports - match cfg only by hostname
+        image_netloc = image_reference.split('/')[0]
+        image_host = image_netloc.split(':')[0]
+
+        for netloc, auth_dict in auths.items():
+            host = netloc.split(':')[0]
+            if host == image_host:
+                break
+        else:
+            if not absent_ok:
+                raise ValueError(f'no auth-cfg found in {docker_cfg=} for {image_reference=}')
```

Comment:
Maybe change the error messages here and the two above a little bit to indicate more precisely what lead to the error, e.g. `no _matching_ auth-cfg found...`. Rationale: This will allow to distinguish what was the root cause in the case of an error.